### PR TITLE
Specify selector types

### PR DESCRIPTION
**Pull request by vmx**
Add a description of selector types we want to support. There isn't
a nice syntax yet, but there is a schema of the serialised form of
the selectors.

### CODE REVIEW

File: selectors/selectors.md [hunk 1]
```diff
@@ -110,295 +110,269 @@ These components imply or expand into the following things:
...
-```
+Selects a single CID.
 
-**TODO**: put selectors into the grand magnificent global universal multicodec table...
 
-### Cid Selector
+### Path Selector
 
-This is a basic selector, that means exactly a single CID, and thus a single graph node. Nothing more.
+Selects a single path segment.
```
Comment (momack2) [inline selectors/selectors.md:123]: why rephrase the description to something less descriptive?

Comment (momack2) [reply selectors/selectors.md:123]: old one was "This is a selector w/ a path. Absolute ones must start with a CID"

File: selectors/selectors.md [hunk 2]
```diff
@@ -110,295 +110,269 @@ These components imply or expand into the following things:
...
-```go
-// Binary: <selector-code><length><path-selector-code><binary-cid><binary-path>
-//     or: <path-selector-code><length><binary-path>
-// String: /sel/path/<multibase-encoded-cid>/a/b/c/d/...
-type PathSelector struct {
-  Path Path
-}
-```
 
-### Glob Selector
```
Comment (momack2) [inline selectors/selectors.md]: why the change from glob to array? AFAICT they don't behave exactly the same - are you really proposing a different selector, or just trying to rename the same one initially described here with regexs for paths?

File: selectors/selectors.md [hunk 3]
```diff
@@ -110,295 +110,269 @@ These components imply or expand into the following things:
...
+### Recursive Selector
 
-This is a selector with a [unix shell glob](https://en.wikipedia.org/wiki/Glob_(programming)) style thing. Absolute ones must start with a CID. It supports the following syntax:
+Uses another Selector to select things recursively. It also takes a limit on how often this recursion should happen (the limit can be set to "infinity" to keep traversing until there are no matching paths left.
```
Comment (momack2) [inline selectors/selectors.md:133]: nice description =]

File: selectors/selectors.md (hunk 2 above)
Comment (momack2) [reply selectors/selectors.md]: how are path and array selectors different from each other? what does a "slice" of items mean vs a "path segment" - could be useful to describe these a little more and explain their usage to help resolve questions like this

File: selectors/selectors.md [hunk 4]
```diff
@@ -110,295 +110,269 @@ These components imply or expand into the following things:
...
-```go
-// Binary: <selector-code><length><glob-selector-code><binary-cid><binary-path>
-//     or: <glob-selector-code><length><binary-path>
-// String: /sel/path/<multibase-encoded-cid>/a/*/**
-type GlobSelector struct {
-    Glob Path
-}
-```
 
-### Multi Selector
```
Comment (momack2) [inline selectors/selectors.md]: I found multi selectors confusing, but if "selector queries" are supposed to be a replacement for them we need to enhance this definition. I'd really like to understand what I'd use selector queries for and how they'd work (do they union/intersect? do I get to choose?)

File: selectors/selectors.md [hunk 5]
```diff
@@ -110,295 +110,269 @@ These components imply or expand into the following things:
...
+        "year": 2328
+      },
+      "rank": "vice admiral"
+    }
+  }
+}
 ```
-<sel-sexp>  ::= <varint-2+><operator>(<selector>)+
+
+The final Selector could look like this:
```
Comment (momack2) [inline selectors/selectors.md:168]: I thought selectors were normally described as strings? do you want to include the notation that a client might use to say look up the rank of Katherin?

File: selectors/selectors.md [hunk 6]
```diff
@@ -110,295 +110,269 @@ These components imply or expand into the following things:
...
+      {"selectPath": "birthday"},
+      {"selectPath": "year"}
+    ]
+  }
+}
 ```
 
-Option 2: more complicated at the `sel-sexp` level, but avoids having to embed the multiselector preamble.
 
+### Getting a certain number of parent blocks in a blockchain
```
Comment (momack2) [inline selectors/selectors.md:185]: 👍

File: selectors/selectors.md [hunk 7]
```diff
@@ -110,295 +110,269 @@ These components imply or expand into the following things:
...
-    return nil
+IPLD Schema
+-----------
+
+```ipldsch
+# This is the main entry point for the current selectors
+type CidRootedSelector struct {
+  root Cid
+  # Each element matches a path segement (or several, in some cases)
+  selectors [Selector]
```
Comment (momack2) [inline selectors/selectors.md:322]: why is this plural?

File: selectors/selectors.md [hunk 8]
```diff
@@ -110,295 +110,269 @@ These components imply or expand into the following things:
...
 
-func (d *traversedDag) GetNode(path Path) Node {
-    if d.s.IncludesPath(path) {
-        return d.d.GetNode(path)
-    }
-    return nil
+# A catch all for all types of selectors. They are split between recursive and
+# non-recursive ones as a recursive selector can't have another recursive
+# selector embedded
+type Selector union {
```
Comment (momack2) [inline selectors/selectors.md:328]: you define selector types here, but not the functions the selector exposes (the string/byte representation, the type, etc) - intentional?

**Review by momack2 — COMMENTED** [review]
Thanks for the writeup, @vmx. I think the selector types could use some more description and have a few questions about title, usage, etc, but appreciate you writing it up!

File: selectors/selectors.md (hunk 7 above)
Comment (vmx) [reply selectors/selectors.md:322]: In this spec, there isn't the idea of "one selector with a certain type" anymore. All selectors can be combined. So if you want to access something with a deeply nested path, it is a list of path selector, as a single path selector is one segment of the path only.

File: selectors/selectors.md (hunk 8 above)
Comment (vmx) [reply selectors/selectors.md:328]: The functionality is described above. This is a description on how they are encoded within the IPLD data model.

**Comment by vmx** [comment]
@momack2 I've rebased this commit. Sorry for the earlier version it was confusing. It wasn't meant to be diffed with the old version. The overlap with it was just coincidentally. This PR should be read as a new document, not as changes to the old ones. I hope things make more sense then.

Hence I just replied to the questions you had within this new context. If you still have questions, feel free to ask.

File: selectors/selectors.md (hunk 5 above)
Comment (vmx) [reply selectors/selectors.md:168]: No, currently there is no simple string encoding, it's always a structure that can e.g. be represented as JSON. So find out her rank would be:

```json
{
  "cidRootedSelector": {
    "root": "cidabcdef",
    "selectors": [
      {"selectPath": "characters"},
      {"selectPath": "kathryn-janeway"},
      {"selectPath": "rank"}
    ]
  }
}
```

Though one future string representation will most likely look like `cidabcdef/characters/kathryn-janeway/rank`.

File: selectors/selectors.md (hunk 7 above)
Comment (warpfork) [reply selectors/selectors.md:322]: also, in case it's that the syntax may not yet be familiar: `[Selector]` means an array of things of type `Selector`.  (It's a syntax roughly inspired from GraphQL schemas.)

File: selectors/selectors.md (hunk 1 above)
Comment (warpfork) [reply selectors/selectors.md:123]: Has to do with how these are now a composable model.

If you want to start at an absolute CID and then navigate to some path within it, you... combine a CID selector with a Path selector.

So the whole "absolute" thing just kinda went away.  Or rather, It's not that it's gone, it's just that it's not a weird special case of this particular feature now.

Comment (warpfork) [reply selectors/selectors.md:123]: (There are some nomenclature issues to be sorted out about how we refer to single selector components vs the whole aggregated selector, to be sure...)

**Review by warpfork — APPROVED** [review]

File: selectors/selectors.md [hunk 9]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+  | SelectArrayPosition "selectArrayPosition"
+  | SelectArraySlice "selectArraySlice"
+  | SelectArrayAll "selectMapAll"
+} representation keyed
+
+# Paths are split into their individual segments
+type PathSegement String
+
+# Selects a specific full path
+type SelectPath PathSegment
```
Comment (hannahhoward) [inline selectors/selectors.md:350]: If this is a full path, should it be `type SelectPath [PathSegment]`?

Comment (hannahhoward) [reply selectors/selectors.md:350]: actually curious cause the spec above says a path selector is only to select a segment.

**Review by hannahhoward — COMMENTED** [review]
Minor nits, LGTM

File: selectors/selectors.md [hunk 10]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+
+In some DAG you want to get one specific value you know the path of. Let's say you want to get the birth year of a specific character of a specific show.
+
+Example data (as JSON) with CID `cidabcdef`. To keep it simple it's represented as a single JSON structure, but imagine e.g. the characters are linked by CID.
+
+```json
+{
+  "show": "start-trek-voyager",
+  "characters": {
+    "kathrin-janeway": {
```
Comment (rvagg) [inline selectors/selectors.md]: *nit: `kathryn-janeway` for correctness and to match the selector just below

File: selectors/selectors.md [hunk 11]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+
+
+### Recursive Selector
+
+Uses another Selector to select things recursively. It also takes a limit on how often this recursion should happen (the limit can be set to "infinity" to keep traversing until there are no matching paths left.
+
+The recursion may also be stopped by a given CID.
+
+
+### Selector Query
```
Comment (rvagg) [inline selectors/selectors.md]: is "query" the best name for this? could it just be "Selector List" or is there context I'm missing (very possible!)?

File: selectors/selectors.md [hunk 12]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+```
+
+It will be a Recursive Selector following a long a certain field (`prev` in this case):
+
+```json
+{
+  "cidRootedSelector": {
+    "root": "cidabcdef",
+    "selectors": [
+      {"selectCid": "cidabcdef"},
```
Comment (rvagg) [inline selectors/selectors.md]: is this necessary since it's the root anyway?

File: selectors/selectors.md [hunk 13]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+```
+
+If it's a file in some directory, you can also start at a deeper level:
+
+```json
+{
+  "cidRootedSelector": {
+    "root": "cidabcdef",
+    "selectors": [
+      {"selectCid": "cidabcdef"},
```
Comment (rvagg) [inline selectors/selectors.md]: same question as above, is this serving an important purpose or demonstrating the possible redundancy?

File: selectors/selectors.md [hunk 14]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+  | SelectRecursive "selectRecursive"
+} representation keyed
+
+# This is a subset of the selectors that can be used within a recursive selector
+type SelectorNonRecursive union {
+  | SelectPath "selectPath"
+  | SelectArrayAll "selectArrayAll"
+  | SelectArrayPosition "selectArrayPosition"
+  | SelectArraySlice "selectArraySlice"
+  | SelectArrayAll "selectMapAll"
```
Comment (rvagg) [inline selectors/selectors.md]: `SelectArrayAll` -> `SelectMapAll`

File: selectors/selectors.md [hunk 15]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+} representation keyed
+
+# Paths are split into their individual segments
+type PathSegement String
+
+# Selects a specific full path
+type SelectPath PathSegment
+
+# Selects all elements of an array, it's similar to a `/*`
+type SelectArrayAll Null
```
Comment (rvagg) [inline selectors/selectors.md:353]: Couldn't this just be rolled up in to one of the other two array types? Maybe position with a `-1` or slice without bounds specified.

File: selectors/selectors.md [hunk 16]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+type ArrayPosition Int
+
+# Selects a slice of items out of an array
+type ArraySlice struct {
+  start optional Int
+  end optional Int
+}
+
+# Selects all keys one level deep, it's similar to a `/*`
+type SelectMapAll Null
```
Comment (rvagg) [inline selectors/selectors.md:365]: There seems to be a linguistic disconnect here between 'path' and 'map', could it be brought in better alignment? `SelectMapAll` -> `SelectPathAll` or `SelectPath` -> `SelectKey`?

File: selectors/selectors.md [hunk 17]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+type PathSegement String
+
+# Selects a specific full path
+type SelectPath PathSegment
+
+# Selects all elements of an array, it's similar to a `/*`
+type SelectArrayAll Null
+
+# Selects a specific item from an array
+type ArrayPosition Int
```
Comment (rvagg) [inline selectors/selectors.md]: `ArrayPosition` -> `SelectArrayPosition`

File: selectors/selectors.md [hunk 18]
```diff
@@ -108,3 +108,271 @@ These components imply or expand into the following things:
...
+
+# Selects a slice of items out of an array
+type ArraySlice struct {
```
Comment (rvagg) [inline selectors/selectors.md]: `ArraySlice` -> `SelectArraySlice`

File: selectors/selectors.md (hunk 15 above)
Comment (rvagg) [reply selectors/selectors.md:353]: Further on the map types: this is likely my ignorance speaking, but is it strictly necessary to treat them different to other properties for the purposes of selecting or is it a matter of needing strict typing for values? You could do:

* array positions as paths/keys, just using integers as your selector value
* "all" for both map and array are essentially the same
* slice is distinct to array types

File: selectors/selectors.md (hunk 9 above)
Comment (vmx) [reply selectors/selectors.md:350]: Good catch! The comment is wrong. A full path is formed by the `cidRootedSelector` which has an array of selectors, hence can have an array of `SelectPath`, hence a full path.

File: selectors/selectors.md (hunk 12 above)
Comment (vmx) [reply selectors/selectors.md]: It is not. Thanks for catching, it must be a leftover from an older version.

File: selectors/selectors.md (hunk 15 above)
Comment (vmx) [reply selectors/selectors.md:353]: Having `SelectArrayAll` is more explicit (which I like), it also plays nice with type systems.

Have a separate type for "Map" helps for better error checking/messages. You should normally know if it's an array or a map. Though it might make sense to add a "I don't know/care" in the future if it turns out it makes sense.

File: selectors/selectors.md (hunk 16 above)
Comment (vmx) [reply selectors/selectors.md:365]: Thanks for bringing this up. I see this as a broader issue (even IPLD wide), hence I've opened https://github.com/ipld/specs/issues/103 and will leave it like that for now, so we can get it merged.

File: selectors/selectors.md (hunk 11 above)
Comment (vmx) [reply selectors/selectors.md]: Good catch, this is now the `CidRootedSelector`. That also explains why the examples below were wrong :)

**Comment by vmx** [comment]
Thanks @hannahhoward and @rvagg all issues should be addressed. Please give your thumbs up/approve this PR so that we can merge it soon.

File: selectors/selectors.md (hunk 10 above)
Comment (mikeal) [reply selectors/selectors.md]: best captain, no contest.

don’t @ me :P

**Review by rvagg — APPROVED** [review]

**Review by hannahhoward — APPROVED** [review]

**Comment by whyrusleeping** [comment]
Without getting into the details of the schema for encoding selectors, or their human readable representations, the expressibility of this scheme seems good enough. @jbenet is gonna be upset that selectors arent always rooted in a CID, but I think separating them makes the composition story cleaner. That said, do we have a term that describes 'selector rooted with a CID' that isnt that long? I think the bulk of juans argument for having selectors always be CID rooted is that it provides an easy term to talk about that particular thing.

**Comment by mikeal** [comment]
I vote for “rooted selector” and “un-rooted selector” :)

**Comment by mikeal** [comment]
I’ll also quickly note that having them separated makes for a very nice and more advanced pinning system around selectors, where you create a pin for an “un-rooted selector” and then you simply set and re-set the “root” of the selector.

**Comment by hannahhoward** [comment]
@vmx @momack2 can we go ahead and merge this? who else needs to sign off? (It would be nice to be able to reference the updated file in some stuff that I am doing)

File: selectors/selectors.md [hunk 19]
```diff
@@ -108,3 +108,269 @@ These components imply or expand into the following things:
...
+
+
+### Recursive Selector
+
+Uses another Selector to select things recursively. It also takes a limit on how often this recursion should happen (the limit can be set to "infinity" to keep traversing until there are no matching paths left.
+
+The recursion may also be stopped by a given CID.
+
+
+### CID Rooted Selector
```
Comment (Stebalien) [inline selectors/selectors.md:138]: Hm. Shouldn't we have two?

1. A CID selector (which we have).
2. A sequential selector that selects each selector one after the other?

File: selectors/selectors.md [hunk 20]
```diff
@@ -108,3 +108,269 @@ These components imply or expand into the following things:
...
+
+
+### Path Selector
+
+Selects a single path segment.
+
+
+### Array Selector
+
+Selects a slice or single item of an array of items.
```
Comment (Stebalien) [inline selectors/selectors.md:128]: Wouldn't the "single item" selector just be the path selector? That is, I assume the array selector will always return an array.

File: selectors/selectors.md (hunk 19 above)
Comment (Stebalien) [reply selectors/selectors.md:138]: Alternatively, we could drop the CID selector and have:

1. A CID rooted selector. `{"root": Cid, selector: ...}`. Use with no selector (or the identity selector?) to select the root cid.
2. A sequential selector.

File: selectors/selectors.md [hunk 21]
```diff
@@ -108,3 +108,269 @@ These components imply or expand into the following things:
...
+
+
+### Recursive Selector
+
+Uses another Selector to select things recursively. It also takes a limit on how often this recursion should happen (the limit can be set to "infinity" to keep traversing until there are no matching paths left.
+
+The recursion may also be stopped by a given CID.
```
Comment (Stebalien) [inline selectors/selectors.md:135]: What if we just had a general purpose "stop" condition selector?

File: selectors/selectors.md (hunk 20 above)
Comment (Stebalien) [reply selectors/selectors.md:128]: Actually, isn't this more of a "range" selector? That is, we're selecting a "range" of keys (from an object or an array).

File: selectors/selectors.md [hunk 22]
```diff
@@ -108,3 +108,269 @@ These components imply or expand into the following things:
...
+
+Selects a single path segment.
+
+
+### Array Selector
+
+Selects a slice or single item of an array of items.
+
+
+### Recursive Selector
```
Comment (Stebalien) [inline selectors/selectors.md:131]: Let's say that I want to lookup a specific path on each block along a chain. I think we'll need a way to fork. Maybe something like:

```
{
  "recursive": {
    "follow": ...,
    "select": ...,
    "stop": ...,
  }
}
```

**Review by Stebalien — COMMENTED** [review]
We should look at other graph query systems (and SQL). We don't have to invent this _entirely_ ourselves and we're definitely not going to make the best thing we can if we try.

File: selectors/selectors.md (hunk 19 above)
Comment (vmx) [reply selectors/selectors.md:138]: I introduced the CID Rooted Selector so that we can have other ones, like a sequential one (i'd call it multi selector) in the future. It can be added to the spec later on. I just wanted to keep things as minimal as possible to make agreement easier.

Comment (vmx) [reply selectors/selectors.md:138]: You can use `{"root": CID, selector: []}` to select only a CID.

File: selectors/selectors.md (hunk 20 above)
Comment (vmx) [reply selectors/selectors.md:128]: That's actually a good point. I named it after the data structures it operates on. But it makes sense to name them after the result (what they actually return). Is it OK if I create a PR once this one is merged (so that we can then discuss it there if anyone has objections, instead of postponing this one even more).

File: selectors/selectors.md (hunk 21 above)
Comment (vmx) [reply selectors/selectors.md:135]: I'll open an issue for that once merged.

File: selectors/selectors.md (hunk 22 above)
Comment (vmx) [reply selectors/selectors.md:131]: In the current spec we don't have any filtering capability at all yet. Currently we just return everythig we traverse. I'll open an issue for that once this spec is merged.

**Comment by vmx** [comment]
I merge this spec now, I think we found agreement on the general direction. @Stebalien has pointed out several things which I'll address in subsequent PRs/issues.

File: selectors/selectors.md (hunk 20 above)
Comment (vmx) [reply selectors/selectors.md:128]: Actually this part is outdated. We don't have a general "Array Selector" anymore. We have a single item selector, slice and all elements selector now. I'll update the spec.

Comment (vmx) [reply selectors/selectors.md:128]: https://github.com/ipld/specs/pull/104

File: selectors/selectors.md (hunk 21 above)
Comment (vmx) [reply selectors/selectors.md:135]: https://github.com/ipld/specs/issues/105

File: selectors/selectors.md (hunk 22 above)
Comment (vmx) [reply selectors/selectors.md:131]: https://github.com/ipld/specs/issues/106

File: selectors/selectors.md (hunk 19 above)
Comment (vmx) [reply selectors/selectors.md:138]: About the sequential selector: https://github.com/ipld/specs/issues/107

Comment (Stebalien) [reply selectors/selectors.md:138]: My point was to keep each selector as minimal as possible. The CID rooted selector does two things: selects a CID and a sequence of follow-on selectors.

Comment (vmx) [reply selectors/selectors.md:138]: Yes. Currently it's the only "entry point". So the CID Rooted Selector is kind of the only proper Selector, all other ones are "Selector Components" or something like that.
